File: 17.Backend Security - Everything You Need to Know/code/typescript/nosql_injection_mongodb_1.ts
// X NEVER, string concatenation
const query = "SELECT * FROM users WHERE email = '" + userInput + "'";

// OK ALWAYS, parameterised ($1 is the slot)
const row = await db.query(
    "SELECT id, name FROM users WHERE email = $1",
    [userInput]  // passed separately, treated purely as data
);

// With an ORM (Prisma), parameterised automatically
const user = await prisma.user.findFirst({ where: { email: userInput } });

File: 17.Backend Security - Everything You Need to Know/code/rust/nosql_injection_mongodb_1.rs
// X NEVER, string concatenation
let query = format!("SELECT * FROM users WHERE email = '{}'", user_input);

// OK ALWAYS, parameterised ($1 is the slot)
let row = sqlx::query!(
    "SELECT id, name FROM users WHERE email = $1",
    user_input // passed separately, treated purely as data
)
.fetch_one(&pool)
.await?;

// With an ORM (SeaORM), parameterised automatically
let user = User::find()
    .filter(user::Column::Email.eq(user_input))
    .one(&db)
    .await?;
